### PR TITLE
Make `allow_*` flags work with caching

### LINKED ISSUE #654: Interaction of cache and `allow_*` flags
Detected by @myrazma 

There are two three major issues with the `allow_*` flags.

### Issue 1
`allow_recommending_already_recommended` completely fails when consecutively requesting the same batch size because of a false positive cache hit.

Minimal reproducing example:
```python
from baybe.campaign import Campaign
from baybe.parameters.numerical import NumericalDiscreteParameter

campaign = Campaign(
    NumericalDiscreteParameter("param", [0, 1]),
    allow_recommending_already_recommended=False,
)

for _ in range(3):
    campaign.recommend(1)   # <-- should run out of options in the third iteration, but doesn't
```

### Issue 2
There is a flaw in the caching logic in presence of pending experiments because returning the cache content requires that the new recommendation is requested with the same pending experiments as used for the cache, which is currently not checked.

### Issue 3
Similarly to issue 2, also the values of the `allow_*` flags must be consistent for both calls as a requirement for returning the cache content.

### PR DESCRIPTION
**Pull request by AdrianSosic**
Fixes #654. 

Improvements that are out of scope for this bugfix PR:
* Less strict cache invalidation: For several cases, the PR clears the cache too conservatively (see also comments added).  In particular, caching is not used at all / immediately cleared as soon as pending experiments enter the stage. By storing the previous recommendation context and using more fine-grained conditions, cache usage could be further extended.
* Instead of manually maintaining the caching logic and conditions, we might also consider switching to a built-in approach (e.g. using `@cached` directly on the recommender level), which would potentially simply the logic significantly and avoid future bugs due to not having to think about edge cases. However, the caching approach would then be even stricter as it is now, since it would only work based on the incoming objects.

### CODE REVIEW

File: tests/test_pending_experiments.py [hunk 1]
```diff
@@ -116,7 +116,7 @@ def test_pending_points(campaign, batch_size, fake_measurements):
     # some recommenders which could also trivially avoid overlap
     with temporary_seed(1337):
         rec1 = campaign.recommend(batch_size)
-    campaign._cached_recommendation = pd.DataFrame()  # ensure no recommendation cache
+    campaign._cached_recommendation is None
```
Comment (Scienfitz) [inline tests/test_pending_experiments.py]:
```suggestion
    campaign.clear_cache()
```

[the remark this answers is not in the document]
Comment (Scienfitz) [reply tests/test_pending_experiments.py]: should be an assignment not an assert

File: baybe/campaign.py [hunk 2]
```diff
@@ -346,6 +367,9 @@ def update_measurements(
             ValueError: If the given data contains indices not present in existing
                 measurements.
         """
+        # With new measurements, the recommendations must always be recomputed
+        self.clear_cache()
```
Comment (Scienfitz) [inline baybe/campaign.py]: wouldnt you rather put this below all the validations instead of the top of this method?

File: baybe/campaign.py [hunk 3]
```diff
@@ -295,8 +316,8 @@ def add_measurements(
             numerical_measurements_must_be_within_tolerance: Flag indicating if
                 numerical parameters need to be within their tolerances.
         """
-        # Invalidate recommendation cache first (in case of uncaught exceptions below)
-        self._cached_recommendation = pd.DataFrame()
+        # With new measurements, the recommendations must always be recomputed
+        self.clear_cache()
```
Comment (Scienfitz) [inline baybe/campaign.py]: should we move this below the validations?

File: baybe/campaign.py [hunk 4]
```diff
@@ -346,6 +367,9 @@ def update_measurements(
             ValueError: If the given data contains indices not present in existing
                 measurements.
         """
+        # With new measurements, the recommendations must always be recomputed
```
Comment (Scienfitz) [inline baybe/campaign.py]:
```suggestion
        # With changed measurements, the recommendations must always be recomputed
```

File: baybe/campaign.py [hunk 5]
```diff
@@ -460,25 +489,40 @@ def recommend(
...
+        # IMPROVE: The cache handling needs improvement:
+        #   * Currently, we simply invalidate the cache whenever pending experiments are
+        #     provided, because in order to use it, we need to check if the previous
+        #     call was done with the same pending experiments.
+        #   * Ideally, once the previous pending experiments are stored,
+        #     we could do a smart check if the cache can be reused even when the
+        #     new pending experiments differ. For example, the cache is still valid
+        #     if a previous call was done without pending experiments and the new
+        #     pending experiments have no intersection with the cached recommendations.
+        #     Additional shortcuts might be possible.
```
Comment (Scienfitz) [inline baybe/campaign.py]: is this correct? Arent pending experiments entering via fantasy? So I'd expect pending_experiments to have an influence ont he recommendations even if they have no exact overlap

File: baybe/campaign.py [hunk 6]
```diff
@@ -460,25 +489,40 @@ def recommend(
...
             raise ValueError(
                 f"You must at least request one recommendation per batch, but provided "
                 f"{batch_size=}."
             )
 
-        # Invalidate cached recommendation if pending experiments are provided
-        if (pending_experiments is not None) and not pending_experiments.empty:
-            self._cached_recommendation = pd.DataFrame()
+        if pending_experiments is not None:
+            self.clear_cache()  # see IMPROVE comment above
```
Comment (Scienfitz) [inline baybe/campaign.py]: why is the note not here in this place?

File: tests/test_campaign.py [hunk 7]
```diff
@@ -283,3 +409,35 @@ def test_acquisition_value_computation(ongoing_campaign: Campaign):
...
+        param("allow_recommending_pending_experiments", False, id="pending"),
+    ],
+)
+@pytest.mark.parametrize("change", [True, False], ids=["change", "no_change"])
+def test_cache_invalidation(
+    campaign: Campaign, attribute: str, value: Any, change: bool
+):
+    """Altering mutable public attributes invalidates the cache."""
+    if isinstance(value, bool):
+        new_value = not value if change else value
```
Comment (Scienfitz) [inline tests/test_campaign.py]:
```suggestion
        new_value = value ^ change
```

File: baybe/campaign.py [hunk 8]
```diff
@@ -80,6 +82,13 @@ def default_allow_flag(campaign: Campaign) -> bool | UnspecifiedType:
     return default_allow_flag
 
 
+def _set_with_cache_cleared(instance: Campaign, attribute: Attribute, value: _T) -> _T:
+    """Attrs-compatible hook to clear the cache when changing an attribute."""
+    if value != getattr(instance, attribute.name):
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/campaign.py:87]: No, I specifically want eq-comparison

File: tests/test_campaign.py (hunk 7 above)
Comment (AdrianSosic) [reply tests/test_campaign.py]: Much better, indeed 😅  thx

File: baybe/campaign.py (hunk 2 above)
Comment (AdrianSosic) [reply baybe/campaign.py]: yes, probably better, done

File: baybe/campaign.py (hunk 3 above)
Comment (AdrianSosic) [reply baybe/campaign.py]: yes, probably better, done

File: baybe/campaign.py (hunk 6 above)
Comment (AdrianSosic) [reply baybe/campaign.py]: because there is a second place below where it's referenced. But it's fine, I moved it to this place

File: baybe/campaign.py (hunk 5 above)
Comment (AdrianSosic) [reply baybe/campaign.py]: Thx for the catch! What I wrote is obviously bullshit! I think a actually had a very different situation in mind but wrote down the wrong explanation – unfortunately, I don't remember what it was. But maybe this already shows that we probably don't want to overdue it with cache optimization. Simpler is probably just more robust 😬 I remember this famous quote about cache invalidation ... 😅

File: baybe/campaign.py [hunk 9]
```diff
@@ -275,6 +288,14 @@ def validate_config(cls, config_json: str) -> None:
         config = json.loads(config_json)
         _validation_converter.structure(config, Campaign)
 
+    def _cache_recommendation(self, df: pd.DataFrame, /) -> None:
+        """Cache the given recommendation."""
+        self._cached_recommendation = df.copy()
```
Comment (AVHopp) [inline baybe/campaign.py:293]: This removes any already existing cached recommendations - if this is intended, please simply resolve, just want to make sure.

File: baybe/campaign.py [hunk 10]
```diff
@@ -378,7 +402,6 @@ def update_measurements(
 
         # Reset fit number and cached recommendations
```
Comment (AVHopp) [inline baybe/campaign.py]: Outdated comment as you do not delete the cache anymore.

**Review by AVHopp — APPROVED** [review]

File: baybe/campaign.py (hunk 9 above)
Comment (AdrianSosic) [reply baybe/campaign.py:293]: Yeah, I think this is fine 👍🏼

File: baybe/campaign.py (hunk 10 above)
Comment (AdrianSosic) [reply baybe/campaign.py]: thx, fixed
